update gguf backend to use Chat-completion API

falkbene opened the pull request:
The response structure for the `logprobs` of the /completion API was changed here: https://github.com/ggml-org/llama.cpp/commit/57bb2c40cd94c5a09f5210ed8264cc93b21c4b7e. Furthermore, /completion API is now `Legacy` ( https://platform.openai.com/docs/guides/completions). This commit adapts the gguf backend to utilize the /chat/completion API and now handles the `logprobs` response correctly. Moreover, this resolves an issue https://github.com/ggml-org/llama.cpp/issues/12591 where the `llama-server` did not recognize the `echo` parameter, as it is no longer necessary.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=2856) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=2856) before we can accept your contribution.<br/><hr/>**Benedikt Falk** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/EleutherAI/lm-evaluation-harness?pullRequest=2856) it.</sub>

baberabb commented:
Hi! We should still keep the completions api as long as GGUF is supporting it. Otherwise will have to chat format the prompt for base models as well.

falkbene commented:
There is an issue with the current implementation as I pointed out in the issue mentioned in my last message. Starting a llama-server with the newest llama.cpp does not support the echo parameter anymore which is accessed in the lm-eval gguf file that I modified. Furthermore, the response structure of the logprobs that is expected in lm_eval/models/gguf.py was also changed in an update of Llama.cpp ( see last comment). So the current gguf implementation of LM-Evaluation-Harness  throws errors when I use it. My edits should fix that at least for the gguf file, we could also use the completions API, but need to adapt the expected response structure.

falkbene commented:
I think we could still use the completions API, but still have to adapt to the response coming from the server, as the response structure has changed.

teekuningas commented:
Without _echo_, does it not return probs for only the continuation after _prompt + continuation_? That is e.g for

"What is the capital of Japan? Tokyo" the continuation might be ".". 

You could use grammar constraints. So the prompt would be "What is the capital of Japan?", you would constraint the response to be e.g " Tokyo." and get correct probs without _echo_.
